algosdk/testing/teal_blackbox.py

```diff
@@ -0,0 +1,720 @@
+                mode, key
+            ), f"each key must be a DryrunAssertionTypes appropriate to {mode}. This is not the case for key {key}"
+
+    return inputs, assertions
+
+
+class SequenceAssertion:
+    def __init__(
+        self,
+        predicate: Union[Dict[Tuple, Union[str, int]], Callable],
+        enforce: bool = False,
+        name: str = None,
+    ):
+        self.definition = predicate
+        self.predicate, self._expected = self.prepare_predicate(predicate)
+        self.enforce = enforce
+        self.name = name
+
+    def __repr__(self):
+        return f"SequenceAssertion({self.definition})"[:100]
+
+    def __call__(
+        self, args: list, actual: Union[str, int]
+    ) -> Tuple[bool, str]:
+        assertion = self.predicate(args, actual)
+        msg = ""
+        if not assertion:
+            msg = f"SequenceAssertion for '{self.name}' failed for for args {args}: actual is [{actual}] BUT expected [{self.expected(args)}]"
+            if self.enforce:
+                assert assertion, msg
+
+        return assertion, msg
+
+    def expected(self, args: list) -> Union[str, int]:
+        return self._expected(args)
+
+    @classmethod
+    def prepare_predicate(cls, predicate):
+        if isinstance(predicate, dict):
+            return (
+                lambda args, actual: predicate[args] == actual,
+                lambda args: predicate[args],
+            )
+
+        if not isinstance(predicate, Callable):
+            # constant function in this case:
+            return lambda _, actual: predicate == actual, lambda _: predicate
+
+        try:
+            sig = signature(predicate)
+        except Exception as e:
+            raise Exception(
+                f"callable predicate {predicate} must have a signature"
+            ) from e
+
+        N = len(sig.parameters)
+        assert N in (1, 2), f"predicate has the wrong number of paramters {N}"
+
+        if N == 2:
+            return predicate, lambda _: predicate
+
+        # N == 1:
+        return lambda args, actual: predicate(
+            args
+        ) == actual, lambda args: predicate(args)
+
+
+def mode_has_assertion(
+    mode: ExecutionMode, assertion_type: DryRunAssertionType
+) -> bool:
+    missing = {
+        ExecutionMode.Signature: {
+            DryRunAssertionType.cost,
+            DryRunAssertionType.lastLog,
+        },
+        ExecutionMode.Application: set(),
+    }
+    if assertion_type in missing[mode]:
+        return False
+
+    return True
+
+
+def dig_actual(
+    dryrun_resp: dict,
+    assert_type: DryRunAssertionType,
+    assertion_arg: Any = None,
+) -> Union[str, int, bool]:
+    txns = dryrun_resp["txns"]
+    assert (
+        len(txns) == 1
+    ), f"expecting exactly 1 transaction but got {len(txns)}"
+    txn = txns[0]
+    mode = (
+        ExecutionMode.Signature
+        if "logic-sig-messages" in txn
+        else ExecutionMode.Application
+    )
+    is_app = mode == ExecutionMode.Application
```

Comment:
@tzaffi Friendly pointer to what appears to be a duplicate `is_app` definition.  It's defined again several lines below within the same lexical scope.  I _assume_ 1 of 2 the definitions can be safely removed.

```suggestion

```